activiti-core/activiti-api-impl/activiti-api-process-runtime-impl/src/main/java/org/activiti/runtime/api/impl/ExtensionsVariablesMappingProvider.java

```diff
@@ -166,15 +166,26 @@ private Map<String, Object> calculateOutPutVariables(MappingExecutionContext map
             }
         }
 
-        Map<String, Object> result = expressionResolver.resolveExpressionsMap(
-            new SimpleMapExpressionEvaluator(availableVariables), outboundVariables);
+        return resolveExpressions(mappingExecutionContext, availableVariables, outboundVariables);
+    }
 
+    private Map<String, Object> resolveExpressions(MappingExecutionContext mappingExecutionContext,
+                                                   Map<String, Object> availableVariables,
+                                                   Map<String, Object> outboundVariables) {
         if (mappingExecutionContext.hasExecution()) {
-            result = expressionResolver.resolveExpressionsMap(
-                new VariableScopeExpressionEvaluator(mappingExecutionContext.getExecution()), result);
+            if (availableVariables != null && !availableVariables.isEmpty()) {
+                return expressionResolver.resolveExpressionsMap(
+                    new CompositeVariableExpressionEvaluator(
+                        new SimpleMapExpressionEvaluator(availableVariables),
+                        new VariableScopeExpressionEvaluator(mappingExecutionContext.getExecution())),
+                        outboundVariables);
+            }
+            return expressionResolver.resolveExpressionsMap(
+                new VariableScopeExpressionEvaluator(mappingExecutionContext.getExecution()), outboundVariables);
+        } else {
+            return expressionResolver.resolveExpressionsMap(
+                new SimpleMapExpressionEvaluator(availableVariables), outboundVariables);
         }
```

Comment:
Maybe we can extract some method here to make it more readable?

Reply:
I moved execution expression resolving to a different method https://github.com/Activiti/Activiti/pull/4295/commits/263b1c4808f2a17db3b0b7cd2878fb1ef69ac9d9